### PR TITLE
2.3.2 wpf

### PR DESCRIPTION
**Pull request by Atreides07**
### Description of Change ###

Preview of Xamarin.Forms.WPF Renderer project.

### API Changes ###

Added:
 - Converters
 - ActivityIndicatorRenderer (ProgressBar)
 - ButtonRenderer (Button)
 - BoxViewRenderer (Rectangle)
 - CellControl (in current version only ViewCell)
 - DatePickerRenderer (DatePickerRenderer)
 - EditorRenderer (TextBox)
 - EntryRenderer (FormsTextBox)
 - FrameRenderer (Border)
 - ImageRenderer (Image)
 - ImageRenderer (TextBox)
 - ListViewRenderer (ListView)
 - PageRenderer (Panel)
 - PickerRenderer
 - ProgressBarRenderer (ProgressBar)
 - ScrollViewRenderer (ScrollViewer)
 - SearchBarRenderer (TextBox)
 - SliderRenderer (Slider)
 - StepperRenderer
 - SwitchRenderer
 - TableViewRenderer
Changed:
   i tried to keep original Xamarin.Forms.Core project as is to ease migration to new versions of Xamarin Forms. But in TemplatedItemsList i change two methods:
               void IList.Clear()
		{
			throw new NotSupportedException();
		}

		bool IList.Contains(object item)
		{
			throw new NotImplementedException();
		}
  changed to 
                void IList.Clear()
		{
			ListProxy.Clear();
		}

		bool IList.Contains(object item)
		{
		    return ListProxy.Contains(item);
		}

and in AssemblyInfo include VisibilityTo Xamarin.Forms.Platform.WPF
[assembly: InternalsVisibleTo("Xamarin.Forms.Platform.WPF")]

### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [ ] Rebased on top of master at time of PR
- [ ] Changes adhere to coding standard
- [ ] Consolidate commits as makes sense

### CODE REVIEW

**Comment by jassmith** [comment]
Hey can you drop me an email: jas at microsoft dot com

**Comment by Atreides07** [comment]
@jassmith Yesterday I wrote a letter to your address from sherievakhmed at live.ru . Did you receive it?

**Comment by Atreides07** [comment]
@jassmith  how can I get a new branch instead of merging into an existing (2.3.2) branch?

File: Xamarin.Forms.Platform.WPF/Forms.cs [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+			s_isInitialized = true;
+		}
+
+		class DefaultApplication : Application
+		{
+		}
+
+		internal class WPFDeviceInfo : DeviceInfo
+		{
+			internal const string BWPorientationChangedName = "Xamarin.WPF.OrientationChanged";
```
Comment (ahmedalejo) [inline Xamarin.Forms.Platform.WPF/Forms.cs]: I guess you meant WpfOrientationChangedName

**Comment by ahmedalejo** [comment]
@Atreides07 this must have been a heck of an adventure. 👍 
@jassmith will certainly fill you on about xamarin forms roadmap and if there´s an ongoing internal effort to accomplish this.

Great initiative mate

**Comment by rmarinho** [comment]
Hey, can you do this PR agains the branch wpf ? thanks, we will also need you to sign the CLA agreement.

**Comment by Atreides07** [comment]
Probably better to do reject this PR. There i made a lot of mistakes because it was in a hurry when I created this project. I'm working on a new version now. I do PR with new project and i going to migrate code from this project to the new version after review.

**Comment by delfuria** [comment]
Hi, do you have any estimated released timeframe for this PR (or for a new one) ?
I would like to use the WPF support in a large project :-)
Thanks

**Comment by Atreides07** [comment]
@delfuria Right now I work hard to make the new version more quickly. Unfortunately, it will take at least 2-4 months. I do not recommend to use this version becouse so many not made as I wanted. New version will support version 2.3.4. A nice bonus: In the new version you can use the default interface and can be able to using Mushup as plugin (http://mahapps.com/) for modern metro design becouse default desing of WPF controls don't look modern.

**Comment by brianlagunas** [comment]
Not to burst anyone's bubble, but I think trying to have XF support WPF is a very bad idea.  Of course, that's just my opinion.

**Comment by Atreides07** [comment]
A year ago, I, too, was sure that Win7 does not need support. Myself, as a developer, much more like the UWP platform. The first time I tried to convince the owners that the support for Windows 7 is not needed. But our app designed for tablets and desktops has a lot of users on the Windows platform and is very much on Win7. The first time we had a separate version of the UI for Win7 and separately for the XF. And common code since ViewModels. I am very tired ensure that my team is doing the right implementation for Win7 and XF. I have implemented support for WPF and brought my team that would  simplify their work. In any case, my project XF.WPF saved a lot of time on my team because our application is growing rapidly and is often changed to the new requirements. 

Also i wanted to make support XF Mac before the team Xamarin announced this. But our potential users of Mac applications is less than 2% since in our country are not popular Mac in schools.

In the future, after the completion of work on this project and I want to try to make XF Web. Not for production. Just for fun

**Comment by brianlagunas** [comment]
As an enterprise architect, I would never allow a WPF app to be written with XF in my company.  But, like I said, that's just me.  If you find value in it, great.  I personally will never recommend any one build a WPF app with XF.  I strongly feel that this should **not** be added to the XF repo.  I think Xamarin should spend their time on more important things.  They should stabilize what they have now, and greatly improve performance before trying to attempt to add another platform.  In my mind, this is not a priority that should have any time spent on it.

**Comment by rmarinho** [comment]
I don't agree with you here @brianlagunas XF is a cross platform solution, if one is going to use XF for WPF is because he's already using Forms for other platform.. 
As a architect  you know there are timelines / resources available to conclude projects, the idea is with this backend you can now extend your reach for your XF app to Windows7 computers. Also we are looking at this support to be added by someone outside of our team, parallel to what we are doing.

**Comment by Atreides07** [comment]
@brianlagunas I also have several years of experience in application development for enterpise, where large projects are developed a large number of developers for several years. At a previous job we had a team only on Android team consisted of 20 developers. Xamarin, in principle, was not considered because it was very important to get the best possible performance. The project was a lot of C ++ code. And very surprised that you use XF for the enterprise. Perhaps it suits you.
I am now in a very small start-up company. I have a total of 6 developers including myself. In addition, sometimes we have to do tasks for backend. And XF with WPF solves our problems perfectly.
And I am very surprised by your indignation WPF support in the XF. Any middle developer knows about loosely coupled architecture. If you look at the code you can see that I did not touch Xamarin.Forms.Core general. I needed access to internal objects, and so I just add one line:
[Assembly: InternalsVisibleTo ( "Xamarin.Forms.Platform.WPF")]
I support WPF since version 2.1.0 the XF and is regularly updated to the new version alone. Xamarin team does not know about the existence of my project at all.
If you do not need XF WPF simply do not use it. For you, nothing will change. XF will remain as it is.
As a good architect, you should understand these simple principles of software development.

P.S. I also began work on an alternative Xamarin Forms. If I succeed something that I hope to realize something before the end of the year. While it is very early for the production but shows significantly better performance than the Xamarin Forms. But consumes more memory.

**Comment by brianlagunas** [comment]
@rmarinho you don't have to agree. I could argue with anyone all day about this. I know I'll never use it :)

I would rather Xamarin focus on higher priorities. I don't see a massive call to action asking for this coming from customers or the community.

**Comment by opcodewriter** [comment]
I completely agree with @brianlagunas 

>  I think Xamarin should spend their time on more important things. They should stabilize what they have now, and greatly improve performance before trying to attempt to add another platform. In my mind, right now this is not a priority and no time should be spent on it.

Xamarin Forms has still a lot of issues when running on Android and iOS. There are bugs in the layout engine, depending on the platform sometimes it behaves differently. I'm talking about basic controls like Grid and StackLayout. Sometimes these do not perform as expected.  And beside bugs, I wish small but important features from Windows platform weren't overlooked. 
Xamarin Forms should be rock solid on Android and iOS but it isn't. In a couple of months Xamarin Forms is going to be **3 years** old.  

I also don't think it's a good idea to mix this kind of framework for mobile with one for desktop. One example: each have their own patterns in terms of navigation.

**Comment by opcodewriter** [comment]
I completely agree with @brianlagunas 

>  I think Xamarin should spend their time on more important things. They should stabilize what they have now, and greatly improve performance before trying to attempt to add another platform. In my mind, right now this is not a priority and no time should be spent on it.

Xamarin Forms has still a lot of issues when running on Android and iOS, bugs and performance issues. There are bugs in the layout engine, depending on the platform sometimes it behaves differently. I'm talking about basic controls like Grid and StackLayout. Sometimes these do not perform as expected.  And beside bugs, I wish small but important features from Windows platform weren't overlooked. 
Xamarin Forms should be rock solid on Android and iOS but it isn't. In a couple of months Xamarin Forms is going to be **3 years** old.  

The Xamarin Forms team is such a small team. Currently Xamarin Forms is trying to reach iOS, Android, Windows UWP, Mac OSX, WinRT, WinPhone8 and now WPF? Sorry but to me me, Xamarin Forms is a Jack of all trades, master of none.
And each year, there's a new Android and iOS version which needs to tested and maintained...
Anyone who's serious in developing quality apps understands how difficult this is.
And Visual Studio 2017 is being released in a couple of months!!!  What about XAML Previewer, does it even work OK, on both VS2015 and VS2017?

I also don't think it's a good idea to mix this kind of framework for mobile with one for desktop. One example: each have their own patterns in terms of navigation. 

By the way, I still have no idea who leads Xamarin Forms. I thought it's Jason Smith but honestly, I don't think that's the case. I only see him merging stuff in main branch and closing things

**Comment by Atreides07** [comment]
@opcodewriter I work with Xamarin a very long time. Even before the Xamarin name. When announced the first version of the XF all sounded exactly the same arguments. XF is not necessary. In Xamarin so a lot of bugs to fix first and then make Xamarin Forms. Each platform has its own its own UI and navigation and therefore can not be wasting time on Xamarin Forms. Nevertheless Xamarin Forms received recognition from many developers.

Nevertheless, I understand you. Despite the fact that we are living with a year XF WPF I was sure that no one except us, this project is not needed. A month ago I did not think about what to do PR until I saw a tweet from Miguel. It turns out many have supported demand for WPF

**Comment by opcodewriter** [comment]
I'm not questioning if Xamarin Forms is needed.  And I was only referring to the differences between Desktop and Mobile. 
My main point was Xamarin Forms team is already a small team. If someone from XF has time to merge, test and sign-off changes on a new platform, then great. To me it appeared there's little time, that's why things are progressing so slowly for the last 3 years.

**Comment by ahmedalejo** [comment]
@Atreides07 it´s a pity people frown on this. Especially because this wasn´t event built by the xamarin forms team. But just a suggestion:

keep doing what works best for you. For example @brianlagunas created or is advertising Prism for xamarin.forms, being one of the the top notch feature is **URI based Navigation** (:thinking: isn´t that a web thing.... not!) he invested time in something he believed was worthwhile and hopefully there are others who believe the same, 

Like you said about Xamarin.Forms inception. Everyone was skeptic, probably even those frowning on Xamarin.Forms.WPF now, didn´t think it was what the shot(like: Xamarin.Forms! oh no! 😨 is that Windows Forms grandson?).

Bu i bet my cents that even Xamarin took a while to convince itself about Xamarin.Forms **potential** and it was spooled several times.

So @Atreides07, 👍 for your effort, unconditional sharing is one of the pillars of **OpenSource** initiative.

**Comment by opcodewriter** [comment]
@Atreides07 I apologize for being maybe too vocal. I've nothing against your work and PR. I think everybody appreciates your efforts. Thank you.

**Comment by Atreides07** [comment]
@ahmedalejo @opcodewriter Thanks for the nice reviews. I remade the work from scratch where I finished a lot of things that were not originally. On Tuesday I will make PR in the WPF thread. Now I will make PR with version 2.3.2. Within a week I'll try to make version 2.3.5.

**Pull request closed without merging**